src/main/java/org/snomed/snowstorm/extension/ExtensionAdditionalLanguageRefsetUpgradeService.java

```diff
@@ -154,67 +151,52 @@ private Map<Long, ReferenceSetMember> getReferencedComponents(String branchPath,
 	private List<ReferenceSetMember> addOrUpdateLanguageRefsetComponents(String branchPath, AdditionalRefsetExecutionConfig config, Map<Long, ReferenceSetMember> languageRefsetMembersToCopy) {
 		BranchCriteria branchCriteria = versionControlHelper.getBranchCriteria(branchPath);
 		List<ReferenceSetMember> result = new ArrayList<>();
-		// batch here as the max allowed in terms query is 65536
 
-		// Get all concepts that need changing from ToCopy members
-		Set<String> conceptIds = new HashSet<>();
-		for (List<Long> batch : partition(languageRefsetMembersToCopy.keySet(), 10_000)) {
-			NativeSearchQueryBuilder queryBuilder = new NativeSearchQueryBuilder()
-					.withQuery(boolQuery().must(branchCriteria.getEntityBranchCriteria(Description.class)))
-					.withFilter(termsQuery(DESCRIPTION_ID, batch))
-					.withPageable(LARGE_PAGE);
-
-			try (final SearchHitsIterator<Description> searchHitsIterator = elasticsearchTemplate.searchForStream(queryBuilder.build(), Description.class)) {
-				searchHitsIterator.forEachRemaining(hit -> conceptIds.add(hit.getContent().getConceptId()));
-			}
-		}
+		// Collect all concepts that need changing from ToCopy members
+		Set<String> conceptIds = new HashSet<>(languageRefsetMembersToCopy.values().stream().map(ReferenceSetMember:: getConceptId).collect(Collectors.toSet()));
 
-		// Get all relevant descriptions
+		// Collect all relevant description IDs
 		Set<Long> descriptionIds = new HashSet<>();
-		Set<Long> extensionDescriptionIds = new HashSet<>();
-		Map<Long, Set<Description>> conceptToDescriptionsMap = new Long2ObjectOpenHashMap<>();
+		descriptionIds.addAll(languageRefsetMembersToCopy.values().stream().map(item -> Long.valueOf(item.getReferencedComponentId())).collect(Collectors.toSet()));
+
+		// Get all existing language reference set members
+		Map<String, ReferenceSetMember> existingLanguageRefsetMembersToUpdate = new HashMap<>();
 		for (List<String> batch : partition(conceptIds, 10_000)) {
 			NativeSearchQueryBuilder queryBuilder = new NativeSearchQueryBuilder()
-					.withQuery(boolQuery().must(branchCriteria.getEntityBranchCriteria(Description.class)))
-					.withFilter(termsQuery(CONCEPT_ID, batch))
+					.withQuery(boolQuery()
+							.must(branchCriteria.getEntityBranchCriteria(ReferenceSetMember.class))
+							.must(termQuery(REFSET_ID, config.getDefaultEnglishLanguageRefsetId()))
+							.must(termsQuery(ACCEPTABILITY_ID_FIELD_PATH, Concepts.PREFERRED, Concepts.ACCEPTABLE))
+							.must(termsQuery(CONCEPT_ID, batch))
+					).withFields(REFERENCED_COMPONENT_ID, ACTIVE, ACCEPTABILITY_ID_FIELD_PATH)
```

Comment:
@QuyenLy87 You can simplify the logic even further by filtering out PTs not in the copy list from last step. The results will indicate whether there is a new PT added in the extension branch already for a given concept. This then can be used to decide whether to add/update language members from the copy list.

Reply:
I think we should keep this logic here. We have to get all the existing language refset members for given conceptIds because we need to know the existing ones which can be used for updating if needed. If you only get the new PTs added in the extension branch, then we will miss the existing ones for updating. @CoderMChu